File: Android/app/src/main/java/ru/inwords/inwords/data/source/webService/WebRequestsManagerImpl.java
package ru.inwords.inwords.data.source.webService;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;
import ru.inwords.inwords.core.util.Function;
import ru.inwords.inwords.core.util.SchedulersFacade;
import ru.inwords.inwords.data.dto.EntityIdentificator;
import ru.inwords.inwords.data.dto.User;
import ru.inwords.inwords.data.dto.UserCredentials;
import ru.inwords.inwords.data.dto.WordTranslation;
import ru.inwords.inwords.data.dto.game.Game;
import ru.inwords.inwords.data.dto.game.GameInfo;
import ru.inwords.inwords.data.dto.game.GameLevel;
import ru.inwords.inwords.data.dto.game.LevelScore;
import ru.inwords.inwords.data.dto.game.LevelScoreRequest;
import ru.inwords.inwords.data.source.webService.session.AuthInfo;
import ru.inwords.inwords.data.source.webService.session.AuthInfoKt;
import ru.inwords.inwords.data.source.webService.session.SessionHelper;
import ru.inwords.inwords.data.source.webService.session.TokenResponse;
import ru.inwords.inwords.data.sync.PullWordsAnswer;

public class WebRequestsManagerImpl implements WebRequestsManager {
    private final WebApiService apiService;
    private final SessionHelper sessionHelper;

    private final AuthInfo authInfo;

    @Inject
    WebRequestsManagerImpl(WebApiService apiService, BasicAuthenticator authenticator, SessionHelper sessionHelper, AuthInfo authInfo) {
        this.apiService = apiService;
        this.sessionHelper = sessionHelper;
        this.authInfo = authInfo;

        authenticator.setOnUnauthorisedCallback(() -> getToken().blockingGet()); //TODO COSTIL
    }

    private Single<TokenResponse> setAuthToken(TokenResponse tokenResponse) {
        return Single.fromCallable(() -> {
            this.authInfo.setTokenResponse(tokenResponse);
            return tokenResponse;
        });
    }

    private Single<UserCredentials> setCredentials(UserCredentials userCredentials) {
        return Single.fromCallable(() ->
        {
            authInfo.setCredentials(userCredentials);
            return authInfo.getCredentials();
        });
    }

    private Single<UserCredentials> getCredentials() {
        return Single.fromCallable(authInfo::getCredentials);
    }

    private Single<TokenResponse> getToken() {
        return getCredentials()
                .filter(AuthInfoKt::validCredentials)
                .toSingle()
                .flatMap(s -> applyAuthSessionHelper(apiService.getToken(s)))
                .subscribeOn(SchedulersFacade.io());
    }

    @Override
    public Single<TokenResponse> getToken(UserCredentials userCredentials) {
        return setCredentials(userCredentials)
                .filter(AuthInfoKt::validCredentials)
                .toSingle()
                .flatMap(s -> applyAuthSessionHelper(apiService.getToken(s)))
                .subscribeOn(SchedulersFacade.io());
    }

    @Override
    public Single<TokenResponse> registerUser(UserCredentials userCredentials) {
        return applyAuthSessionHelper(apiService.registerUser(userCredentials))
                .zipWith(setCredentials(userCredentials), (tokenResponse, u) -> tokenResponse)
                .subscribeOn(Schedulers.io());
    }

    @Override
    public Single<String> getLogin() {
        return applySessionHelper(apiService::getLogin)
                .subscribeOn(SchedulersFacade.io());
    }

    @Override
    public Single<User> getAuthorisedUser() {
        return applySessionHelper(apiService::getAuthorisedUser)
                //.flatMap(Observable::fromIterable)
                .subscribeOn(SchedulersFacade.io());
    }

    @Override
    public Single<User> getUserById(int id) {
        return applySessionHelper(b -> apiService.getUserById(b, id))
                //.flatMap(Observable::fromIterable)
                .subscribeOn(SchedulersFacade.io());
    }

    @Override
    public Single<List<EntityIdentificator>> insertAllWords(List<WordTranslation> wordTranslations) {
        return applySessionHelper(b -> apiService.addPairs(b, wordTranslations))
                .subscribeOn(SchedulersFacade.io());
    }

    @Override
    public Single<Integer> removeAllServerIds(List<Integer> serverIds) {
        return applySessionHelper(b -> apiService.deletePairs(b, serverIds))
                .subscribeOn(SchedulersFacade.io());
    }

    @Override
    public Single<PullWordsAnswer> pullWords(List<Integer> serverIds) {
        return applySessionHelper(b -> apiService.pullWordsPairs(b, serverIds))
                .subscribeOn(SchedulersFacade.io());
    }

    @Override
    public Single<List<GameInfo>> getGameInfos() {
        return applySessionHelper(apiService::getGameInfos)
                .subscribeOn(SchedulersFacade.io());
    }

    @Override
    public Single<Game> getGame(int gameId) {
        return applySessionHelper(b -> apiService.getGame(b, gameId))
                .subscribeOn(SchedulersFacade.io());
    }

    @Override
    public Single<GameLevel> getLevel(int levelId) {
        return applySessionHelper(b -> apiService.getLevel(b, levelId))
                .subscribeOn(SchedulersFacade.io());
    }

    @Override
    public Single<LevelScore> getScore(LevelScoreRequest levelScoreRequest) {
        return applySessionHelper(b -> apiService.getGameScore(b, levelScoreRequest))
                .subscribeOn(SchedulersFacade.io());
    }

    @Override
    public Single<Boolean> uploadScore(List<LevelScoreRequest> levelScoreRequests) {
        return applySessionHelper(b -> apiService.uploadScore(b, levelScoreRequests).toSingleDefault(true))
                .subscribeOn(SchedulersFacade.io());
    }

    private String getBearer() {
        return authInfo.getTokenResponse().getBearer();
    }

    private <R> Single<R> applySessionHelper(Function<String, Single<R>> func) {
        return sessionHelper
                .requireThreshold()
                .andThen(Single.defer(() -> func.apply(getBearer())))
                .doOnError(throwable -> sessionHelper.interceptError(throwable).blockingAwait());
    }

    private Single<TokenResponse> applyAuthSessionHelper(Single<? extends TokenResponse> query) {
        return sessionHelper
                .resetThreshold()
                .andThen(query)
                .onErrorResumeNext(throwable -> sessionHelper.interceptError(throwable)
                        .andThen(Single.error(() -> { //error here
                            setAuthToken(TokenResponse.errorToken());
                            return throwable;
                        })))
                .flatMap(this::setAuthToken);
    }
}

File: Android/app/src/main/java/ru/inwords/inwords/data/source/webService/WebRequestsManagerImpl.kt
package ru.inwords.inwords.data.source.webService

import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.functions.BiFunction
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.BehaviorSubject
import ru.inwords.inwords.core.ObservableTransformers
import ru.inwords.inwords.core.util.SchedulersFacade
import ru.inwords.inwords.data.dto.EntityIdentificator
import ru.inwords.inwords.data.dto.User
import ru.inwords.inwords.data.dto.UserCredentials
import ru.inwords.inwords.data.dto.WordTranslation
import ru.inwords.inwords.data.dto.game.*
import ru.inwords.inwords.data.dto.google.AudioConfig
import ru.inwords.inwords.data.dto.google.Input
import ru.inwords.inwords.data.dto.google.TtsSynthesizeRequest
import ru.inwords.inwords.data.dto.google.Voice
import ru.inwords.inwords.data.source.webService.session.AuthInfo
import ru.inwords.inwords.data.source.webService.session.SessionHelper
import ru.inwords.inwords.data.source.webService.session.TokenResponse
import ru.inwords.inwords.data.source.webService.session.validCredentials
import ru.inwords.inwords.data.sync.PullWordsAnswer
import javax.inject.Inject

class WebRequestsManagerImpl @Inject
internal constructor(private val apiService: WebApiService,
                     authenticator: BasicAuthenticator,
                     private val sessionHelper: SessionHelper,
                     private val authInfo: AuthInfo) : WebRequestsManager {

    private val authenticatedNotifierSubject = BehaviorSubject.create<Boolean>()

    init {
        authenticator.setOnUnauthorisedCallback(object : BasicAuthenticator.OnUnauthorisedCallback {
            override fun perform(): TokenResponse {
                return getToken().blockingGet()
            }
        }) //TODO COSTIL
    }

    private fun setAuthToken(tokenResponse: TokenResponse): Single<TokenResponse> {
        return Single.fromCallable {
            this.authInfo.tokenResponse = tokenResponse

            authenticatedNotifierSubject.onNext(!authInfo.isNoToken)

            tokenResponse
        }
    }

    private fun Single<UserCredentials>.updateToken(): Single<TokenResponse> {
        return flatMap {
            if (validCredentials(it)) {
                apiService.getToken(it)
            } else {
                setAuthToken(AuthInfo.unauthorisedToken)
            }
        }
                .applyAuthSessionHelper()
                .subscribeOn(SchedulersFacade.io())
    }

    override fun getToken(): Single<TokenResponse> {
        return authInfo.getCredentials().updateToken()
    }

    override fun getToken(userCredentials: UserCredentials): Single<TokenResponse> {
        return authInfo.setCredentials(userCredentials).updateToken()
    }

    override fun getUserEmail(): Single<String> {
        return authInfo.getCredentials().map { it.email }
    }

    override fun registerUser(userCredentials: UserCredentials): Single<TokenResponse> {
        return apiService.registerUser(userCredentials)
                .applyAuthSessionHelper()
                .zipWith(authInfo.setCredentials(userCredentials),
                        BiFunction<TokenResponse, UserCredentials, TokenResponse> { tokenResponse, u -> tokenResponse })
                .subscribeOn(Schedulers.io())
    }

    override fun getLogin(): Single<String> {
        return valve()
                .flatMap { getAuthToken() }
                .flatMap { apiService.getLogin(it) }
                .interceptError()
                .subscribeOn(SchedulersFacade.io())
    }

    override fun getAuthorisedUser(): Single<User> {
        return valve()
                .flatMap { getAuthToken() }
                .flatMap { apiService.getAuthorisedUser(it) }
                .interceptError()
                //.flatMap(Observable::fromIterable)
                .subscribeOn(SchedulersFacade.io())
    }

    override fun getUserById(id: Int): Single<User> {
        return valve()
                .flatMap { getAuthToken() }
                .flatMap { apiService.getUserById(it, id) }
                .interceptError()
                //.flatMap(Observable::fromIterable)
                .subscribeOn(SchedulersFacade.io())
    }

    override fun insertAllWords(wordTranslations: List<WordTranslation>): Single<List<EntityIdentificator>> {
        return valve()
                .flatMap { getAuthToken() }
                .flatMap { apiService.addPairs(it, wordTranslations) }
                .interceptError()
                .subscribeOn(SchedulersFacade.io())
    }

    override fun removeAllServerIds(serverIds: List<Int>): Single<Int> {
        return valve()
                .flatMap { getAuthToken() }
                .flatMap { apiService.deletePairs(it, serverIds) }
                .interceptError()
                .subscribeOn(SchedulersFacade.io())
    }

    override fun pullWords(serverIds: List<Int>): Single<PullWordsAnswer> {
        return valve()
                .flatMap { getAuthToken() }
                .flatMap { apiService.pullWordsPairs(it, serverIds) }
                .interceptError()
                .subscribeOn(SchedulersFacade.io())
    }

    override fun getGameInfos(): Single<List<GameInfo>> {
        return valve()
                .flatMap { getAuthToken() }
                .flatMap { apiService.getGameInfos(it) }
                .interceptError()
                .subscribeOn(SchedulersFacade.io())
    }

    override fun getGame(gameId: Int): Single<Game> {
        return valve()
                .flatMap { getAuthToken() }
                .flatMap { apiService.getGame(it, gameId) }
                .interceptError()
                .subscribeOn(SchedulersFacade.io())
    }

    override fun getLevel(levelId: Int): Single<GameLevel> {
        return valve()
                .flatMap { getAuthToken() }
                .flatMap { apiService.getLevel(it, levelId) }
                .interceptError()
                .subscribeOn(SchedulersFacade.io())
    }

    override fun getScore(levelScoreRequest: LevelScoreRequest): Single<LevelScore> {
        return valve()
                .flatMap { getAuthToken() }
                .flatMap { apiService.getGameScore(it, levelScoreRequest) }
                .interceptError()
                .subscribeOn(SchedulersFacade.io())
    }

    override fun uploadScore(levelScoreRequests: List<LevelScoreRequest>): Single<Boolean> {
        return valve()
                .flatMap { getAuthToken() }
                .flatMap { apiService.uploadScore(it, levelScoreRequests).toSingleDefault(true) }
                .interceptError()
                .subscribeOn(SchedulersFacade.io())
    }

    override fun ttsSynthesize(textToSpeak: String, googleServicesApiKey: String): Single<String> { //TODO
        val request = TtsSynthesizeRequest(
                Input(textToSpeak),
                Voice("en-US", "en-US-Wavenet-D"),
                AudioConfig("OGG_OPUS"))

        return apiService.ttsSynthesize(googleServicesApiKey, request).map { it.audioContent }
    }

    private fun getAuthToken(): Single<String> {
        return authInfo.getCredentials()
                .doOnSuccess {
                    if (!validCredentials(it)) {
                        throw RuntimeException("invalid credentials (no credentials)") //TODO nice exception
                    }
                }
                .map { authInfo.bearer }
    }

    private fun valve(): Single<Unit> {
        return Observable.just(Unit)
                .doOnNext { sessionHelper.requireThreshold() }
                .compose(ObservableTransformers.valve(authenticatedNotifierSubject, false))
                .firstOrError()
    }

    private fun Single<TokenResponse>.applyAuthSessionHelper(): Single<TokenResponse> {
        return doOnSuccess { sessionHelper.resetThreshold() }
                .onErrorResumeNext {
                    if (sessionHelper.interceptAuthError(it)) {
                        setAuthToken(AuthInfo.unauthorisedToken)
                    } else {
                        setAuthToken(AuthInfo.errorToken)
                    }
                }
                .flatMap { setAuthToken(it) }
    }

    private fun <T> Single<T>.interceptError(): Single<T> {
        return doOnError { throwable -> sessionHelper.interceptAuthError(throwable) }
    }
}
